route: Add support for the `route-type`

liangwen12year opened the pull request:
When users have BGP routing setups, it is common practice to blackhole some less-specific routes in order to avoid routing loops, and the BGP router might insert a more specific route dynamically afterwards.

Examples:

```
interfaces:
  - name: eth1
    type: ethernet
    state: up
    ipv4:
      address:
        - ip: 192.0.2.251
          prefix-length: 24
      dhcp: false
      enabled: true

routes:
  config:
    - destination: 198.51.100.0/24
      metric: 150
      next-hop-address: 192.0.2.1
      next-hop-interface: eth1
      table-id: 254
    - destination: 198.51.200.0/24
      route-type: blackhole
```

liangwen12year committed:
route: Add support for the `route-type`

When users have BGP routing setups, it is common practice to blackhole
some less-specific routes in order to avoid routing loops, and the BGP
router might insert a more specific route dynamically afterwards.

Examples:

```
interfaces:
  - name: eth1
    type: ethernet
    state: up
    ipv4:
      address:
        - ip: 192.0.2.251
          prefix-length: 24
      dhcp: false
      enabled: true

routes:
  config:
    - destination: 198.51.100.0/24
      metric: 150
      next-hop-address: 192.0.2.1
      next-hop-interface: eth1
      table-id: 254
    - destination: 198.51.200.0/24
      route-type: blackhole
```

Signed-off-by: Wen Liang <liangwen12year@gmail.com>

liangwen12year committed:
nispor: Drop the filtering for the nispor loopback routes

Loopback is already supported in nmstate, therefore there is no reason
to keep filtering out the nispor loopback routes.

Signed-off-by: Wen Liang <liangwen12year@gmail.com>

cathay4t reviewed (changes requested):
Blackhole route should not hold `next-hop-interface`, we should raise error when user desire so.

For NetworkManager, we always store blackhole route into loopback interface.

liangwen12year commented:
> Blackhole route should not hold `next-hop-interface`, we should raise error when user desire so.
> 
> For NetworkManager, we always store blackhole route into loopback interface.

I dropped the `next-hop-interface` from the blackhole route and added the validation code.

liangwen12year commented on rust/src/lib/route.rs:
```diff
@@ -64,17 +64,43 @@ impl Routes {
+                            || route.next_hop_addr.is_some())
+                    {
+                        return Err(NmstateError::new(
+                            ErrorKind::NotImplementedError,
```
ahh, Maybe I should add test case testing the validation code here.. Will do.

liangwen12year replied:
Tests added.

cathay4t commented on rust/src/lib/nm/nm_dbus/connection/route.rs:
```diff
@@ -113,3 +120,52 @@ pub(crate) fn nm_ip_routes_to_value(
+    Blackhole,
+    Unreachable,
+    Prohibit,
+    Other(u8),
```
We do not need this as `nm_dbus` is not a public API.

cathay4t commented on tests/integration/route_test.py:
```diff
@@ -140,6 +141,67 @@ def test_add_static_route_without_next_hop_address(eth1_up):
+        Route.KEY: {Route.CONFIG: route},
+    }
+
+    with pytest.raises(NmstateNotImplementedError):
```
Should be `NmstateValueError` as this is incorrect desired state.

liangwen12year replied:
I changed into `NmstateValueError`.

cathay4t reviewed (changes requested):
Also need test to make sure verification passed regardless user defined `next-hop-interface: lo` or not.

liangwen12year commented:
> Also need test to make sure verification passed regardless user defined `next-hop-interface: lo` or not.

I added the test which defines  `next-hop-interface: lo` .

cathay4t reviewed (changes requested):
1. I see duplicate route entry in `nmcli c show lo` if I apply the same YAML mulitple times.
2. The kernel and iproute command never show blackhole route as `next-hop-interface: lo`. We should do the same by hiding it for ipv4 route.
3. The `state: absent` remove other routes:

```yml
routes:
  config:
  - state: absent
    route-type: blackhole
```

4. Please convert above as test case(unit or integration).

liangwen12year commented:
> 1. I see duplicate route entry in `nmcli c show lo` if I apply the same YAML mulitple times.
> 2. The kernel and iproute command never show blackhole route as `next-hop-interface: lo`. We should do the same by hiding it for ipv4 route.
> 3. The `state: absent` remove other routes:
> 
> ```yaml
> routes:
>   config:
>   - state: absent
>     route-type: blackhole
> ```
> 
> 4. Please convert above as test case(unit or integration).

Will do, thanks for reviewing it.

liangwen12year commented:
> 1. I see duplicate route entry in `nmcli c show lo` if I apply the same YAML mulitple times.

Actually, regular route has the same issues when applying the same YAML mulitple times (caused by retrieving the routes and showing route table by default ).  e.g. 
```
[root@fcvm rust]# nmcli c show eth1 | grep routes
ipv4.routes:                            { ip = 198.51.100.0/24, nh = 192.0.2.1, mt = 150 }; { ip = 198.51.100.0/24, nh = 192.0.2.1, mt = 150 table=254 }
```

cathay4t commented:
You patch never modify `RouteEntry.sort_key()`, hence dedup never works.

For problem you mentioned, please also fix.

liangwen12year commented:
> You patch never modify `RouteEntry.sort_key()`, hence dedup never works.
> 
> For problem you mentioned, please also fix.

I modified the  `RouteEntry.sort_key()`, dedup should work now, I also added the proper tests for checking dedup.

cathay4t reviewed (approved):
LGTM. Thanks!
